[OpenCL] Provides SYCL kernels for 3D pooling (#97)

lukeiwanski opened the pull request:
* [OpenCL] Adds SYCL kernels for 3D pooling

Uses simple SYCL kernels to provide implementations for all 3D pooling
ops currently in use. These kernels pass the tests, but haven't really
been optimized.

These need benchmarking to compare with Eigen and CPU kernels.

* [OpenCL] Refactors SYCL kernels to use parameter struct

Moves a lot of the functor parameters into a separate data struct, with
the aim of simplifying the functor code.

* [OpenCL] Removes extra fetching of tensor dimensions

We already had the tensor dimensions passed into
LaunchMaxPooling3dGradOP, so don't need to fetch them from the
tensor.

* [OpenCL] Renames SYCL 3D pooling kernels

Adds '3D' to kernel names.

* [OpenCL] Adds 3D pooling SYCL kernel documentation

* [OpenCL] Adds guards around SYCLDevice typedef

* [OpenCL] Use forward input for SYCL MaxPool3DGradGrad

When we had a mix of SYCL and CPU kernels the forward_input would break
and cause computation problems. Now that we have SYCL kernels for all 3D
pooling operations, this is not a problem.

* [OpenCL] Reformats SYCL 3D pooling code

* [OpenCL] Moves SYCL utils into separate header

* [OpenCL] Simplifies SYCL Pool param contructors

Instead of each constructor initialising the data, simplifies the
constructors to call the first constructor.

jwlawson committed:
[OpenCL] Provides SYCL kernels for 3D pooling (#97)

* [OpenCL] Adds SYCL kernels for 3D pooling

Uses simple SYCL kernels to provide implementations for all 3D pooling
ops currently in use. These kernels pass the tests, but haven't really
been optimized.

These need benchmarking to compare with Eigen and CPU kernels.

* [OpenCL] Refactors SYCL kernels to use parameter struct

Moves a lot of the functor parameters into a separate data struct, with
the aim of simplifying the functor code.

* [OpenCL] Removes extra fetching of tensor dimensions

We already had the tensor dimensions passed into
LaunchMaxPooling3dGradOP, so don't need to fetch them from the
tensor.

* [OpenCL] Renames SYCL 3D pooling kernels

Adds '3D' to kernel names.

* [OpenCL] Adds 3D pooling SYCL kernel documentation

* [OpenCL] Adds guards around SYCLDevice typedef

* [OpenCL] Use forward input for SYCL MaxPool3DGradGrad

When we had a mix of SYCL and CPU kernels the forward_input would break
and cause computation problems. Now that we have SYCL kernels for all 3D
pooling operations, this is not a problem.

* [OpenCL] Reformats SYCL 3D pooling code

* [OpenCL] Moves SYCL utils into separate header

* [OpenCL] Simplifies SYCL Pool param contructors

Instead of each constructor initialising the data, simplifies the
constructors to call the first constructor.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

drpngx commented:
Jenkins, test this please.

drpngx commented on tensorflow/core/kernels/pooling_ops_3d.cc:
```diff
@@ -837,6 +845,698 @@ TF_CALL_float(REGISTER_GPU_KERNELS) TF_CALL_half(REGISTER_GPU_KERNELS)
 
 #endif  // GOOGLE_CUDA
 
+#ifdef TENSORFLOW_USE_SYCL
```
This file is starting to get long. I wonder if we should split it. @benoitsteiner WDYT?

benoitsteiner replied:
It would make sense to split the file.

lukeiwanski replied:
How about a separate PR for SYCL3DPoolParams?

lukeiwanski replied:
Fixed in 4e2e36b

benoitsteiner commented on tensorflow/core/util/sycl_util.h:
```diff
@@ -0,0 +1,98 @@
+// TODO(jwlawson): Remove once we have different type accessors for SYCL buffers
+// Providing a way to cast the types of buffers or accessors has been proposed
+// as a SYCL extension, so once this is available we can use an atomic
+// accessor and remove this.
```
The sycl spec states that "The cl::sycl::atomic<T> must be used for elements of a buffer to provide safe atomic access to the buffer from device code." This seems to violate the spec. Why not use cl::sycl::atomic<T> in the max pooling code instead ?

DuncanMcBain replied:
Hi Benoit,
unfortunately there is no way to access elements of an accessor atomically unless all elements are atomic, which is expressed by using a cl::sycl::accessor<T, dims, access::mode::atomic>. Because of this restriction, John has instead used the OpenCL function directly, which is exactly what the SYCL implementation should be doing. It is perfectly valid to call OpenCL functions from SYCL code provided that the host compiler doesn't see them as they will be undefined, which I am told is the case here. cl::sycl::atomic is not user-constructible, and so there's no way to use that directly. Instead it is easiest to use the SPIR-mangled function as John has here.

benoitsteiner replied:
Access to the underlying buffers is requested as follow:
  auto input_access = input_buffer.template get_access<cl::sycl::access::mode::read>(cgh);
  auto output_access = output_buffer.template get_access<cl::sycl::access::mode::write>(cgh);
Can you explain why this can't be modifed into:
  auto input_access = input_buffer.template get_access<cl::sycl::access::mode::atomic>(cgh);
  auto output_access = output_buffer.template get_access<cl::sycl::access::mode::atomic>(cgh);

DuncanMcBain replied:
In particular, you can't turn those into atomic accessors because there is no atomic add for floats in OpenCL 1.2. Even OpenCL 2.0, which expanded the atomic operations available, doesn't have them.

In addition, the underlying buffer type is of int8_t, which has no atomic support either. Similarly, the cl::sycl::atomic type is not user-constructible, but even if it were, it would effectively be doing this anyway (or rather would call this SPIR-mangled function on device, and the C++11 version on host). That said, we'll take note of this feedback - seeing real-world usage is always useful.

Personally, I'm more of a fan of the less-abstract form of atomics where you simply operate atomically on addresses. though that's somewhat lower-level than SYCL aims for. I liked the atomic_view (later atomic_ref) proposals for C++, but I don't believe they made it into C++17.

keryell replied:
@DuncanMcBain this is useful feedback you should send to SYCL committee about what would be helpful to add for this.

Commit:
removing the address space.

Commit:
Compilation error fix and clang format applied

lukeiwanski commented:
@mehdi-goli should be now in CLA

lukeiwanski commented:
Closing. We will open new PR with implementation that doesn't use atomics.

jwlawson committed:
[OpenCL] Provides atomic free MaxPool3DGrad

Atomic support in SYCL is not designed in a way that plays nicely with
Tensorflow and Eigen. Here we provide a new implementation for
MaxPool3DGrad which does not rely on atomics, and so avoids any such
problems.

jwlawson committed:
[OpenCL] Removes unneeded SYCL atomic functions

jwlawson committed:
[OpenCL] Tidies SYCL MaxPool3DGrad kernels

Some tidying up and also adds a local accumulator value which will be
written to memory at the end of the kernel, to decrease the number og
memory writes in the kernel.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

lukeiwanski commented:
@benoitsteiner could you have a look at this updated version that does not contain atomics

vrv commented:
ping for @benoitsteiner

benoitsteiner reviewed (changes requested):
Please move the code to a new file as previously requested to keep it manageable.

Commit:
Acts on feedback: https://github.com/tensorflow/tensorflow/pull/10978#pullrequestreview-53271721

Commit:
Merge branch 'master' into opencl/pooling_ops_3d

lukeiwanski commented:
@benoitsteiner is that better?

benoitsteiner commented:
Jenkins, test this please.
